modules/EmailMan/EmailMan.php

```diff
@@ -1003,6 +1003,179 @@ public function mark_deleted($id)
     }
 
+    /**
+     *
+     * @global array $sugar_config
+     * @param string $module
+     * @param string $uid
+     * @return boolean
+     */
+    public function addOptInEmailToEmailQueue($module, $uid) {
+
+        global $sugar_config;
+
+        $confirmOptInEnabled = isset($sugar_config['email_enable_confirm_opt_in']) && $sugar_config['email_enable_confirm_opt_in'];
+
+        if (!$confirmOptInEnabled) {
+            $this->warn('Confirm Opt In disabled');
+            return false;
+        }
+
+        $relatedBean = BeanFactory::getBean($module, $uid);
+
+        $this->related_type = $relatedBean->module_dir;
+        $this->related_id = $relatedBean->id;
+        $this->related_confirm_opt_in = true;
+
+        $ret = $this->save();
+        return $ret;
+    }
+
+    /**
+     *
+     * @global array $sugar_config
+     * @param EmailAddress $emailAddress
+     * @param string $type
+     * @param string $id
+     * @return boolean
+     * @throws Exception
+     */
+    public function sendOptInEmail(EmailAddress $emailAddress, $type, $id) {
+
+        global $sugar_config;
+
+        $confirmOptInEnabled = isset($sugar_config['email_enable_confirm_opt_in']) && $sugar_config['email_enable_confirm_opt_in'];
+
+        if (!$confirmOptInEnabled) {
+            $this->warn('Confirm Opt In disabled');
+            return false;
+        }
+
+        $guid = $emailAddress->id;
+        if(!$guid) {
+            $focus = BeanFactory::getBean($type, $id);
+            if($focus) {
+                $address = $emailAddress->getPrimaryAddress($focus, $focus->id);
+                $guid = $emailAddress->getEmailGUID($address);
+                $emailAddress->retrieve($guid);
+            } else {
+                $log->error('Incorrect bean');
+                return false;
+            }
+        }
+
+        $focus = BeanFactory::getBean($type, $id);
+        if(!$focus) {
+            throw new Exception('Email address has not related bean.');
+        }
+
+        $ret = $this->sendOptInEmailViaMailer($focus, $emailAddress);
+        return $ret;
+    }
+
+    /**
+     *
+     * @global LoggerManager $log
+     * @global array $sugar_config
+     * @param SugarBean $focus
+     * @param EmailAddress $emailAddress
+     * @return boolean
+     */
+    private function sendOptInEmailViaMailer(
+        SugarBean $focus,
+        EmailAddress $emailAddress
+    ) {
+
+        global $log,  $sugar_config;
+
+        $confirmOptInEnabled = isset($sugar_config['email_enable_confirm_opt_in']) && $sugar_config['email_enable_confirm_opt_in'];
+
+        if (!$confirmOptInEnabled) {
+            $this->warn('Confirm Opt In disabled');
+            return false;
+        }
+
+        $ret = true;
+
+        $emailTemplate = new EmailTemplate();
+
+        $confirmOptInTemplateId =
+                isset($sugar_config['aop']['confirm_opt_in_template_id']) ?
```

Comment:
Note: (or TODO): conceptually it isn't the best practice to take this email template into the AOP config. We should investigate it.